aexpect/ops.py

```diff
@@ -0,0 +1,479 @@
+    if output:
+        output = output.strip()
+    if status != 0 or not output:
+        raise RuntimeError(f'Could not hash {filename} using {cmd}: {output}')
+
+    # parse output
+    hash_str = output.split(maxsplit=1)[0].lower()
+
+    # check length and all chars are hex
+    if expect_len and len(hash_str) != expect_len:
+        raise RuntimeError(f'Resulting hash string has unexpected length '
+                           f'{len(hash_str)}: {hash_str}')
+    if hash_str.strip('0123456789abcdef'):
+        raise RuntimeError('Resulting hash string has unexpected characters: '
+                           + hash_str)
+    return hash_str
+
+
+def extract_tarball(session, tarball, target_dir):
+    """
+    Extract tarball to given dir from a session.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str tarball: full path of tar file that should be extracted
+    :param str target_dir: name of directory where tarball should be extracted
+    :raises: :py:class:`RuntimeError` if tar command returned non-null
+    """
+    cmd = f'tar -C {quote(target_dir)} --strip-components=1 -xapf {quote(tarball)}'
+    status, output = session.cmd_status_output(cmd)
+    if output:
+        output = output.strip()
+    if status != 0:
+        raise RuntimeError(f'Failed to extract {tarball} to {target_dir}: {output}')
+
+
+def ls(session, dir_name):  # pylint: disable=C0103
+    """
+    Run `ls` in given directory through a session.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str dir_name: name of directory
+    :returns: names of files in dir (without path components)
+    :rtype: [str]
+
+    Just like :py:func:`os.listdir`, does not include file names starting with
+    dot (`'.'`)
+    """
+    cmd = f'ls -1UNq {quote(dir_name)}'
+    status, output = session.cmd_status_output(cmd)
+    if output:
+        output = output.strip()
+    if status == 2:      # probably just nothing found
+        return []
+    if status != 0:
+        raise RuntimeError(f'Failed to ls {dir_name}: {output}')
+    return output.splitlines()
+
+
+def glob(session, glob_pattern):
+    """
+    Find files matching given pattern through a session.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str glob_pattern: pattern of filenames where `*`, `?` and such match
+                             any string or not-empty-string and so on.
+    :returns: file names matching pattern, including path (like glob.glob)
+    :rtype: [str]
+    :raises: :py:class:`RuntimeError` if tar ls-command returned non-null and
+             non-2 status
+
+    Much like :py:func:`glob.glob` but implemented using `ls`.
+
+    The glob matching is done locally, not with `ls` since quoting also makes
+    the glob characters '*' and '?' regular characters, so are not interpreted
+    any more in shell.
```

Comment:
This is not really explicit but it won't won't allow globs like `/*/foo` to get all `foo` files from any root directory. So, provided the implementation stays, it'd be better to explicitly expect `glob(session, path, file_glob)` to avoid surprises.

I think it'd be better to rather implement `raw_ls(session, raw_pattern, args='-1UNq')` which would execute the ls without quotation, therefore allowing the server-side globs.

If you still think the client-side filtering is needed, consider the use of `fnmatch`. I think it would be more standard way (although then we could simply drop the whole function as it'd result in `return [_ for _ in ls(sessionos.path.join(path) if fnmatch(_, file_glob)]`)

Reply:
Inspected the function in detail now and frankly, it seems completely redundant. Relying on the server side glob is also more reliable and reproducible when it comes to debugging so I will simply add `is_glob` argument to the directory listing function.